### PR TITLE
Adding consumer tests

### PR DESCRIPTION
**Pull request by rerpha**
added some `KafkaW::Consumer` tests and added dependency injection for it so mock consumers can be used to test calls etc.

### CODE REVIEW

File: src/tests/listener_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+
+class FakePollStatus : public KafkaW::PollStatus {
+public:
+  std::unique_ptr<KafkaW::AbstractMsg> is_Msg() override {
+    FakeMsg mesg("test");
+    return make_unique<FakeMsg>(mesg);
+  };
+//  static FakePollStatus make_Msg(std::unique_ptr<KafkaW::Msg> x)=default;
```
Comment (rerpha) [inline src/tests/listener_tests.cpp]: remove redundant code

File: src/Main.cxx [hunk 2]
```diff
@@ -113,6 +110,18 @@ Main::Main(MainOpt &opt)
     status_producer_topic = ::make_unique<KafkaW::ProducerTopic>(status_producer, main_opt.status_uri.topic);
   }
 }
+void Main::SetUpListener() {
+  KafkaW::BrokerOpt bopt;
```
Comment (rerpha) [inline src/Main.cxx]: refactor names

File: src/KafkaW.h [hunk 3]
```diff
@@ -44,15 +44,25 @@ class TopicOpt {
...
+  virtual uint32_t size()=0;
+  void *kmsg = nullptr;
+  virtual char const *topic_name()=0;
+  virtual int32_t offset()=0;
+  virtual int32_t partition()=0;
+};
+
+class Msg : public AbstractMsg {
 public:
   ~Msg();
```
Comment (rerpha) [inline src/KafkaW.h]: clang-format this

File: src/KafkaW.cxx [hunk 4]
```diff
@@ -186,34 +186,23 @@ bool PollStatus::is_EOP() { return state == -2; }
...
     data = nullptr;
     return ret;
   }
   return nullptr;
 }
 
 Consumer::Consumer(BrokerOpt opt) : opt(opt) {
-  // on_rebalance_start = nullptr;
-  // on_rebalance_assign = nullptr;
+  on_rebalance_assign = {};
```
Comment (rerpha) [inline src/KafkaW.cxx]: check usage of these two variables

File: src/Config.cxx [hunk 5]
```diff
@@ -3,23 +3,15 @@
 #include <condition_variable>
 #include <memory>
 #include <mutex>
+#include <iostream>
```
Comment (rerpha) [inline src/Config.cxx]: check why this was added

File: src/Config.h [hunk 6]
```diff
@@ -2,11 +2,13 @@
 
 #include "KafkaW.h"
 #include "uri.h"
+#include "helper.h"
 #include <atomic>
 #include <chrono>
 #include <memory>
 #include <string>
 #include <vector>
+#include <condition_variable>
```
Comment (matthew-d-jones) [inline src/Config.h]: You've added an include for `condition_variable` here so you should remove it in the `.cxx` file.

File: src/Config.cxx [hunk 7]
```diff
@@ -29,9 +20,6 @@ Listener::Listener(KafkaW::BrokerOpt bopt, uri::URI uri) {
         }
         impl->cv.notify_all();
       };
-  consumer.on_rebalance_assign = {};
-  consumer.on_rebalance_start = {};
-  consumer.add_topic(uri.topic);
 }
 
 Listener::~Listener() {}
```
Comment (matthew-d-jones) [inline src/Config.cxx]: This line can be removed, put `~Listener() = default;` in the header file instead.

File: src/Config.h [hunk 8]
```diff
@@ -23,11 +25,16 @@ class Callback {
   virtual void operator()(string const &msg) = 0;
 };
 
-struct Listener_impl;
+struct Listener_impl{
+  std::unique_ptr<KafkaW::BaseConsumer> consumer;
+  std::mutex mx;
+  std::condition_variable cv;
+  int connected = 0;
```
Comment (matthew-d-jones) [inline src/Config.h]: Change `connected` to a `bool`, this allows some simplification in the `.cxx` file too.

File: src/tests/listener_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    std::unique_ptr<FakeMsg> m2(new FakeMsg);
+    m2->kmsg = (void *)"hello";
+    return FakePollStatus::make_Msg(std::move(m2));
+  }
+};
+
+TEST(kafkaw_consumer_tests, test_config_polling_returns_msg) {
+  std::string expected = "test";
+  auto consumer_ptr = make_unique<MockConsumer>();
+  BrightnESS::ForwardEpicsToKafka::Config::Listener listener1(
```
Comment (matthew-d-jones) [inline src/tests/listener_tests.cpp]: `listener1`->`listener`

File: src/tests/listener_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+  char const *topic_name() override { return "test_topic_name"; };
+  int32_t offset() override { return 1; };
+  int32_t partition() override { return 1; };
+};
+
+class FakePollStatus : public KafkaW::PollStatus {
+public:
+  std::unique_ptr<KafkaW::AbstractMsg> is_Msg() override {
+    FakeMsg mesg("test");
```
Comment (matthew-d-jones) [inline src/tests/listener_tests.cpp]: `mesg` -> `message`

File: src/tests/tests.cxx [hunk 11]
```diff
@@ -11,12 +11,12 @@ MainOpt *Tests::main_opt = nullptr;
...
-    return 1;
-  }
-  auto opt = std::move(po.second);
-  BrightnESS::ForwardEpicsToKafka::tests::Tests::main_opt = opt.get();
+//  auto po = BrightnESS::ForwardEpicsToKafka::parse_opt(argc, argv);
+//  if (po.first) {
+//    return 1;
+//  }
+//  auto opt = std::move(po.second);
+//  BrightnESS::ForwardEpicsToKafka::tests::Tests::main_opt = opt.get();
```
Comment (matthew-d-jones) [inline src/tests/tests.cxx]: Pushing this change is a mistake.

**Comment by rerpha** [comment]
This is now ready for review.

**Pull request merged**